### ISSUE TITLE
Issues with resolving the crc.testing addresses on macOS: server on port 53 misbehaving

### ISSUE THREAD

**Issue by ckvtvm-arch** [code-ready/crc#1675]
`Internal error occurred: registry.redhat.io/rh-sso-7/sso74-openshift-rhel8:latest: Get https://registry.redhat.io/v2/: dial tcp: lookup registry.redhat.io on 172.25.0.10:53: server misbehaving`

**Comment by ckvtvm-arch** [comment]
Internal error occurred: docker.io/jboss/keycloak:latest: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 172.25.0.10:53: server misbehaving

**Comment by ckvtvm-arch** [comment]
I tried all versions of crc starting with 4.6 , 4.4 and now i am with 4.5.1 in my mac.  I have researched around all networking issues from named server, proxy etc, but no luck.  I have added the pull secrets for red hat and docker.  please help me with pointers

**Comment by gbraad** [comment]
> `dial tcp: lookup registry.redhat.io on 172.25.0.10:53: server misbehaving`
> `dial tcp: lookup registry-1.docker.io on 172.25.0.10:53: server misbehaving`

This might indicate a DNS issue.

**Comment by ckvtvm-arch** [comment]
how to resolve this .. i have tried many ways to changing the named server, downgrading the crc, proxy , network settings in mac etc.. no luck...  please let me know how to solve this

**Comment by cfergeau** [comment]
- start with crc 1.18 with a 4.6.1 bundle
- please provide the full logs as suggested by the 'new issue' template (roughly, run `crc delete; crc cleanup; crc setup --log-level debug; crc start --log-level debug`

**Comment by ckvtvm-arch** [comment]
[crc-logs.txt](https://github.com/code-ready/crc/files/5549558/crc-logs.txt)

**Comment by ckvtvm-arch** [comment]
attached the complete log

**Comment by cfergeau** [comment]
Cluster appears to be working reasonably well, I would expect `oc login -u kubeadmin -p Ddbvk-odz4g-NghN8-TiFC8 https://api.crc.testing:6443` and `oc get co` to work fine?

**Comment by ckvtvm-arch** [comment]
You can now run 'crc console' and use these credentials to access the OpenShift web console.
af71194@LC02CHB58MD6R ~ % oc login -u kubeadmin -p Ddbvk-odz4g-NghN8-TiFC8 https://api.crc.testing:6443
error: dial tcp 192.168.64.2:6443: i/o timeout - verify you have provided the correct host and port and that the server is currently running.
af71194@LC02CHB58MD6R ~ % oc get co
Unable to connect to the server: dial tcp 192.168.64.2:6443: i/o timeout
af71194@LC02CHB58MD6R ~ % crc console
Opening the OpenShift Web Console in the default browser...%                                                                                                                                                                                                af71194@LC02CHB58MD6R ~ %

**Comment by cfergeau** [comment]
Does `ssh core@api.crc.testing` give you a ssh password prompt?

**Comment by ckvtvm-arch** [comment]
af71194@LC02CHB58MD6R ~ % ssh core@api.crc.testing
ssh: connect to host api.crc.testing port 22: Operation timed out

**Comment by cfergeau** [comment]
Hmm but the logs show that this was worknig fine during `crc start`? And what you pasted previously shows that this resolves correctly to `192.168.64.2`. So not sure why this would suddenly become unreachable. Is the `hyperkit` process still running? Is there a VPN/firewall/... running?

**Comment by ckvtvm-arch** [comment]
sorry , i was in office vpn.. disconnected and now it works. but my problem is with deploying images in cluster

**Comment by ckvtvm-arch** [comment]
You can now run 'crc console' and use these credentials to access the OpenShift web console.
af71194@LC02CHB58MD6R ~ % oc login -u kubeadmin -p Ddbvk-odz4g-NghN8-TiFC8 https://api.crc.testing:6443
Login successful.

You have access to 58 projects, the list has been suppressed. You can list all projects with 'oc projects'

Using project "default".
af71194@LC02CHB58MD6R ~ % crc console                                                                  
Opening the OpenShift Web Console in the default browser...%                                                                                                                                                                                                af71194@LC02CHB58MD6R ~ % oc get co
NAME                                       VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE
authentication                             4.6.1     False       False         False      12s
cloud-credential                           4.6.1     True        False         False      22d
cluster-autoscaler                         4.6.1     True        False         False      22d
config-operator                            4.6.1     True        False         False      22d
console                                    4.6.1     True        False         False      22d
csi-snapshot-controller                    4.6.1     True        False         False      22d
dns                                        4.6.1     True        False         False      8h
etcd                                       4.6.1     True        False         False      22d
image-registry                             4.6.1     True        False         False      12h
ingress                                    4.6.1     True        False         False      22d
insights                                   4.6.1     True        False         False      22d
kube-apiserver                             4.6.1     True        False         False      22d
kube-controller-manager                    4.6.1     True        False         False      22d
kube-scheduler                             4.6.1     True        False         False      22d
kube-storage-version-migrator              4.6.1     True        False         False      22d
machine-api                                4.6.1     True        False         False      22d
machine-approver                           4.6.1     True        False         False      22d
machine-config                             4.6.1     True        False         False      22d
marketplace                                4.6.1     True        False         False      175m
monitoring                                 4.6.1     True        False         False      22d
network                                    4.6.1     True        False         False      22d
node-tuning                                4.6.1     True        False         False      8h
openshift-apiserver                        4.6.1     True        False         False      145m
openshift-controller-manager               4.6.1     True        False         False      22d
openshift-samples                          4.6.1     True        False         False      12h
operator-lifecycle-manager                 4.6.1     True        False         False      22d
operator-lifecycle-manager-catalog         4.6.1     True        False         False      22d
operator-lifecycle-manager-packageserver   4.6.1     True        False         False      149m
service-ca                                 4.6.1     True        False         False      22d
storage                                    4.6.1     True        False         False      22d

**Comment by ckvtvm-arch** [comment]
Internal error occurred: docker.io/jboss/keycloak:latest: Get "https://registry-1.docker.io/v2/": dial tcp: lookup registry-1.docker.io on 172.25.0.10:53: server misbehaving

**Comment by cfergeau** [comment]
I would suggest to recreate first the cluster with the VPN not running.

**Comment by ckvtvm-arch** [comment]
i think it was how it was created , let me try once again and share the results

**Comment by ckvtvm-arch** [comment]
The authenticity of host 'api.crc.testing (192.168.64.2)' can't be established.
ECDSA key fingerprint is SHA256:YiwBlwQHslwvM5xZ3XwvM6UevFMgyhK7cE+3T33MK2o.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'api.crc.testing' (ECDSA) to the list of known hosts.
core@api.crc.testing: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).

**Comment by ckvtvm-arch** [comment]
but i can login to the console and again deploy image fails for docker, red hat

**Comment by cfergeau** [comment]
One step at a time :) You initially had non-working cluster because of your VPN, now this has improved.

**Comment by cfergeau** [comment]
Can you now give the steps that you are following which are leading you to these `server misbehaving` logs?

**Comment by ckvtvm-arch** [comment]
> One step at a time :) You initially had non-working cluster because of your VPN, now this has improved.

Yes , correct . I was just writing that the original problem is still out there . Also I tried to ssh which has permission denied

**Comment by ckvtvm-arch** [comment]
> Can you now give the steps that you are following which are leading you to these `server misbehaving` logs?

crc start
crc console
login as kubeadmin
select project default or any other project created for our purpose
create secret for dockerhub or redhat as registry.hub.docker.com or registry.redhat.io
select topology 
select container image

populate Image name from external registry as jboss/keycloak or any other image from dockerhub or red hat
validating
Internal error occurred: docker.io/jboss/keycloak:latest: Get "https://registry-1.docker.io/v2/": dial tcp: lookup registry-1.docker.io on 172.25.0.10:53: server misbehaving

**Comment by ckvtvm-arch** [comment]
> The authenticity of host 'api.crc.testing (192.168.64.2)' can't be established.
> ECDSA key fingerprint is SHA256:YiwBlwQHslwvM5xZ3XwvM6UevFMgyhK7cE+3T33MK2o.
> Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
> Warning: Permanently added 'api.crc.testing' (ECDSA) to the list of known hosts.
> [core@api.crc.testing](mailto:core@api.crc.testing): Permission denied (publickey,gssapi-keyex,gssapi-with-mic).

Will this be a problem ?

**Comment by cfergeau** [comment]
Don't pay attention to ssh connectivity, this was only an easy way to check connectivity to your crc instance.

**Comment by ckvtvm-arch** [comment]
> Don't pay attention to ssh connectivity, this was only an easy way to check connectivity to your crc instance.

i have shared the steps, hope that helps

**Comment by ckvtvm-arch** [comment]
the problem is with deploying images from public registry into CRC cluster

**Comment by ckvtvm-arch** [comment]
I will await for your direction.. i have tried all ways for the last few days

**Comment by ckvtvm-arch** [comment]
I will await for your direction.. i have tried all ways for the last few days

**Comment by ckvtvm-arch** [comment]
can we please look into this now

**Comment by ckvtvm-arch** [comment]
i am able to nslookup from command line , but i am not able to pull from CRC console (deploy container image ) 

af71194@LC02CHB58MD6R ~ % nslookup registry.hub.docker.com
Server:		8.8.8.8
Address:	8.8.8.8#53

Non-authoritative answer:
registry.hub.docker.com	canonical name = elb-hub.us-east-1.aws.dckr.io.
elb-hub.us-east-1.aws.dckr.io	canonical name = us-east-1-elbhub-1t5fblb53f6sl-411513349.us-east-1.elb.amazonaws.com.
Name:	us-east-1-elbhub-1t5fblb53f6sl-411513349.us-east-1.elb.amazonaws.com
Address: 52.22.48.121
Name:	us-east-1-elbhub-1t5fblb53f6sl-411513349.us-east-1.elb.amazonaws.com
Address: 3.83.63.98
Name:	us-east-1-elbhub-1t5fblb53f6sl-411513349.us-east-1.elb.amazonaws.com
Address: 54.237.127.59

af71194@LC02CHB58MD6R ~ % nslookup registry.redhat.io     
Server:		8.8.8.8
Address:	8.8.8.8#53

Non-authoritative answer:
registry.redhat.io	canonical name = registry.redhat.io.edgekey.net.
registry.redhat.io.edgekey.net	canonical name = e14353.g.akamaiedge.net.
Name:	e14353.g.akamaiedge.net
Address: 104.104.125.51

**Comment by cfergeau** [comment]
You can get a shell on the CRC VM with `oc debug`, or with `ssh -i ~/.crc/machines/crc/id_rsa core@$(crc ip)`, and check DNS resolutions from the VM

**Comment by ckvtvm-arch** [comment]
af71194@LC02CHB58MD6R ~ % ssh -i ~/.crc/machines/crc/id_rsa core@$(crc ip)
Red Hat Enterprise Linux CoreOS 46.82.202010091720-0
  Part of OpenShift 4.6, RHCOS is a Kubernetes native operating system
  managed by the Machine Config Operator (`clusteroperator/machine-config`).

WARNING: Direct SSH access to machines is not recommended; instead,
make configuration changes via `machineconfig` objects:
  https://docs.openshift.com/container-platform/4.6/architecture/architecture-rhcos.html

---
[systemd]
Failed Units: 1
  systemd-modules-load.service

**Comment by ckvtvm-arch** [comment]
status systemd-modules-load

● systemd-modules-load.service - Load Kernel Modules
Loaded: loaded (/usr/lib/systemd/system/systemd-modules-load.service; static; vendor preset: disabled)
Active: failed (Result: exit-code) since Wed 2020-11-25 23:15:08 UTC; 20min ago
Docs: man:systemd-modules-load.service(8)
man:modules-load.d(5)
Main PID: 866 (code=exited, status=1/FAILURE)
CPU: 12ms

journalctl_pid=866
-- Logs begin at Tue 2020-10-27 01:51:02 UTC, end at Wed 2020-11-25 23:37:28 UTC. --
Nov 25 23:15:08 crc-h66l2-master-0 systemd-modules-load[866]: Failed to insert 'hv_sock': No such device

**Comment by ckvtvm-arch** [comment]
I guess this is something to do with CoreOS. Looks like a vsock is missing in connecting to hypervisor from the host. this may be reason it is failing on connecting to internet. I think there is no issue with DNS as i am able to ping 8.8.8.8

**Comment by gbraad** [comment]
Is the resolver information still matching the VMs IP address?

**Comment by gbraad** [comment]
as per: https://github.com/code-ready/crc/issues/1723#issuecomment-734094132

----

> @ckvtvm-arch can you try this new feature ? https://github.com/code-ready/crc/wiki/VPN-support--with-an--userland-network-stack it should help for your case. Thanks.


----

can you try this and report back?

**Comment by ckvtvm-arch** [comment]
It is waiting on setting up daemon for some time as shown below. Also i havent run crc clean up, hope it wont be a problem as it is a fresh install:

af71194@LC02CHB58MD6R /usr % crc delete
Machine 'crc' does not exist. Use 'crc start' to create it
af71194@LC02CHB58MD6R /usr % vi /etc/hosts
af71194@LC02CHB58MD6R /usr % crc config set network-mode vsock
Network mode changed. Please run `crc cleanup` and `crc setup`.
af71194@LC02CHB58MD6R /usr % crc setup
INFO Checking if podman remote executable is cached 
INFO Checking if goodhosts executable is cached   
INFO Caching goodhosts executable                 
INFO Will use root access: change ownership of /Users/af71194/.crc/bin/goodhosts 
Password:
INFO Will use root access: set suid for /Users/af71194/.crc/bin/goodhosts 
INFO Checking if CRC bundle is cached in '$HOME/.crc' 
INFO Unpacking bundle from the CRC executable     
INFO Checking minimum RAM requirements            
INFO Checking if running as non-root              
INFO Checking if HyperKit is installed            
INFO Setting up virtualization with HyperKit      
INFO Checking if crc-driver-hyperkit is installed 
INFO Installing crc-machine-hyperkit              
INFO Checking file permissions for /etc/hosts     
Setup is complete, you can now run 'crc start' to start the OpenShift cluster
af71194@LC02CHB58MD6R /usr % crc daemon
INFO waiting for clients...                       
INFO listening unix:///Users/af71194/.crc/network.sock

**Comment by ckvtvm-arch** [comment]
Superb... network issue is solved in 4.6.3.  But now i need to resolve the below issue for pulling images from 4.6.3

Created a secret , added image name and validated successfully by enabling the check box for insecure registries from console for container image deploy.

Also linked the secret to default service account

af71194@LC02CHB58MD6R ~ % oc secrets link default dkp --for=pull

Failed to pull image 
"jboss/keycloak@sha256:5dcf63c44980b9d1e9d2b91967a2cd8040176a84a24d41a85affb8cea7f2a310": rpc error: code = Unknown desc = error pinging docker registry registry-1.docker.io: Get "https://registry-1.docker.io/v2/": x509: certificate signed by unknown authority

I need to investigate further on the issue

**Comment by ckvtvm-arch** [comment]
I think i need to replace the default Ingress certificate as detailed in the documentation with one signed by a trusted CA

**Comment by ckvtvm-arch** [comment]
Do i need to set up additional CA certificate as per https://docs.openshift.com/container-platform/4.6/builds/setting-up-trusted-ca.html

**Comment by ckvtvm-arch** [comment]
It didnt worked even after setting up trusted ca as below. generated my own certificate,

> Do i need to set up additional CA certificate as per https://docs.openshift.com/container-platform/4.6/builds/setting-up-trusted-ca.html

i
af71194@LC02CHB58MD6R ssl % oc create configmap registry-cas -n openshift-config \                                                                         
--from-file=registry.hub.docker.com..5000=/private/etc/ssl/cert.pem
configmap/registry-cas created
af71194@LC02CHB58MD6R ssl % oc project
Using project "openshift-config" on server "https://api.crc.testing:6443".
af71194@LC02CHB58MD6R ssl % oc patch image.config.openshift.io/cluster --patch '{"spec":{"additionalTrustedCA":{"name":"registry-cas"}}}' --type=merge
image.config.openshift.io/cluster patched

**Comment by ckvtvm-arch** [comment]
for redhat registry , it successfully pulled image after many attempts by itself . But the router is not working.  

[0m[0m00:42:47,849 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 71) WFLYUT0021: Registered web context: '/auth' for server 'default-server'
[0m[0m00:42:47,942 INFO  [org.jboss.as.server] (ServerService Thread Pool -- 61) WFLYSRV0010: Deployed "keycloak-server.war" (runtime-name : "keycloak-server.war")
[0m[0m00:42:48,018 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0212: Resuming server
[0m[0m00:42:48,026 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: Red Hat Single Sign-On 7.4.3.GA (WildFly Core 10.1.12.SP1-redhat-00001) started in 33570ms - Started 679 of 987 services (701 services are lazy, passive or on-demand)
[0m[0m00:42:48,028 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0060: Http management interface listening on http://127.0.0.1:9990/management
[0m[0m00:42:48,028 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0054: Admin console is not enabled
[0m[33m00:43:01,729 WARN  [org.jgroups.protocols.kubernetes.KUBE_PING] (thread-5,null,t-rhel8-9446f58b7-874tz) failed getting JSON response from Kubernetes Client[masterUrl=https://172.25.0.1:443/api/v1, headers={Authorization=#MASKED:907#}, connectTimeout=5000, readTimeout=30000, operationAttempts=3, operationSleep=1000, streamProvider=org.jgroups.protocols.kubernetes.stream.TokenStreamProvider@374f7d20] for cluster [ejb], namespace [default], labels [null]; encountered [java.lang.Exception: 3 attempt(s) with a 1000ms sleep to execute [OpenStream] failed. Last failure was [java.io.IOException: Server returned HTTP response code: 403 for URL: https://172.25.0.1:443/api/v1/namespaces/default/pods]]

**Comment by ckvtvm-arch** [comment]
I am looking at https://docs.openshift.com/container-platform/4.6/builds/setting-up-trusted-ca.html

**Comment by ckvtvm-arch** [comment]
I may require the public registry certificates

**Comment by guillaumerose** [comment]
Are you on a network with a transparent https proxy ? Do you have custom CAs installed on your machine ?

**Comment by ckvtvm-arch** [comment]
No probs, I resolved the certificate issue by changing the openshift image config since I don't have the public certificate for docker and red hat accessible to me.

Do you know how I can get those public certificates.

If not, we can close this case

**Comment by guillaumerose** [comment]
Red Hat and Docker certificates are all approved by root CAs that are present on default OSes. 
If you can try on a different internet access (4G for instance), I guess it will work.

**Comment by gbraad** [comment]
Custom CA is not necessary for the Red Hat registry. Likely something else wasn't working as expected. Especially as AWS experienced some downtime 3 days ago, this can't be conclusive.

**Comment by ckvtvm-arch** [comment]
I changed the openshift image config to allow insecure registries and it worked for now, but I think 4.6.3 is looking for root certificates to allow connection to Docker and Red Hat registries.
